### ISSUE TITLE
bucket with dot in name fails creation with boto3

### ISSUE THREAD

**Issue by vlcinsky** [spulec/moto#454]
Testing `moto` I have found, that creation of a bucket containing dot in the name fails:

    import pytest
    from moto import mock_s3
    import boto3

    regions = [
        "us-east-1",
        "us-west-2",
        "us-west-1",
        "eu-west-1",
        # "eu-central-1", #ignore, requires @mock_s3bucket_path
        "ap-southeast-1",
        "ap-southeast-2",
        "ap-northeast-1",
        "sa-east-1",
    ]

    @pytest.fixture(params=regions)
    def region_name(request):
        return request.param

    @pytest.fixture
    def s3(region_name):
        return boto3.resource("s3", region_name=region_name)

    bucket_names = ["alpha", "beta.gama"]

    @pytest.fixture(params=bucket_names)
    def bucket_name(request):
        return request.param

    @mock_s3()
    def test_region(s3, bucket_name):
        region_name = s3.meta.client.meta.region_name

        buckcfg = {"LocationConstraint": region_name}
        s3.create_bucket(Bucket=bucket_name, CreateBucketConfiguration=buckcfg)
        buckets = list(s3.buckets.all())
        assert len(buckets) == 1
        assert buckets[0].name == bucket_name
        body_text = "some text"
        key_names = ["hello.txt", "dir/hi.txt", "dir/subdir/good/by.txt"]
        for key_name in key_names:
            s3.Object(bucket_name, key_name).put(Body=body_text)
            body = s3.Object(bucket_name, key_name).get()["Body"].read()
            assert body == body_text

The results are (shortened):

    $ py.test -sv
    ============================================================================================== test session starts ===============================================================================================
    platform linux2 -- Python 2.7.9, pytest-2.8.2, py-1.4.30, pluggy-0.3.1 -- /home/javl/.virtualenvs/boto3/bin/python2
    cachedir: .cache
    rootdir: /home/javl/sandbox/boto3/tests, inifile:
    collected 16 items

    test_s3.py::test_region[alpha-us-east-1] PASSED
    test_s3.py::test_region[alpha-us-west-2] PASSED
    test_s3.py::test_region[alpha-us-west-1] PASSED
    test_s3.py::test_region[alpha-eu-west-1] PASSED
    test_s3.py::test_region[alpha-ap-southeast-1] PASSED
    test_s3.py::test_region[alpha-ap-southeast-2] PASSED
    test_s3.py::test_region[alpha-ap-northeast-1] PASSED
    test_s3.py::test_region[alpha-sa-east-1] PASSED
    test_s3.py::test_region[beta.gama-us-east-1] FAILED
    test_s3.py::test_region[beta.gama-us-west-2] FAILED
    test_s3.py::test_region[beta.gama-us-west-1] FAILED
    test_s3.py::test_region[beta.gama-eu-west-1] FAILED
    test_s3.py::test_region[beta.gama-ap-southeast-1] FAILED
    test_s3.py::test_region[beta.gama-ap-southeast-2] FAILED
    test_s3.py::test_region[beta.gama-ap-northeast-1] FAILED
    test_s3.py::test_region[beta.gama-sa-east-1] FAILED

    ==================================================================================================== FAILURES ====================================================================================================
    ________________________________________________________________________________________ test_region[beta.gama-us-east-1] ________________________________________________________________________________________

    s3 = s3.ServiceResource(), bucket_name = 'beta.gama'

        @mock_s3()
        def test_region(s3, bucket_name):
            region_name = s3.meta.client.meta.region_name

            buckcfg = {"LocationConstraint": region_name}
    >       s3.create_bucket(Bucket=bucket_name, CreateBucketConfiguration=buckcfg)

    test_s3.py:42:
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    ../../../.virtualenvs/boto3/local/lib/python2.7/site-packages/boto3/resources/factory.py:344: in do_action
        response = action(self, *args, **kwargs)
    ../../../.virtualenvs/boto3/local/lib/python2.7/site-packages/boto3/resources/action.py:77: in __call__
        response = getattr(parent.meta.client, operation_name)(**params)
    ../../../.virtualenvs/boto3/local/lib/python2.7/site-packages/botocore/client.py:269: in _api_call
        return self._make_api_call(operation_name, kwargs)
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    self = <botocore.client.S3 object at 0x7fd691c60650>, operation_name = 'CreateBucket', api_params = {'Bucket': 'beta.gama', 'CreateBucketConfiguration': {'LocationConstraint': 'us-east-1'}}

        def _make_api_call(self, operation_name, api_params):
            operation_model = self._service_model.operation_model(operation_name)
            request_dict = self._convert_to_request_dict(
                api_params, operation_model)
            http, parsed_response = self._endpoint.make_request(
                operation_model, request_dict)

            self.meta.events.emit(
                'after-call.{endpoint_prefix}.{operation_name}'.format(
                    endpoint_prefix=self._service_model.endpoint_prefix,
                    operation_name=operation_name),
                http_response=http, parsed=parsed_response,
                model=operation_model
            )

            if http.status_code >= 300:
    >           raise ClientError(parsed_response, operation_name)
    E           ClientError: An error occurred () when calling the CreateBucket operation:

    ../../../.virtualenvs/boto3/local/lib/python2.7/site-packages/botocore/client.py:334: ClientError

It fails the samy way for all regions.

Testing creation of such bucket on real AWS account is working well.

Note, that there is an issue with (old) `boto` regarding similar problem boto/boto#2836, but this is not this case as `boto3` is not having such problem (when not used with `moto`).

**Comment by spulec** [comment]
For bucket names with dots in the name, you will need to use `@mock_s3bucket_path`. I understand that this is not ideal, but it seems that boto changes how it makes requests depending on [whether or not there is a dot in the bucket name](https://github.com/boto/botocore/blob/4c043070a9c7af7e057e6da43ece9641113fa9c8/botocore/utils.py#L624-L646). You might also be able to get away with using both decorators.

The more that these type of issue comes up, I'm tempted to try to merge the decorators or create a combined decorator.

**Comment by vlcinsky** [comment]
I have to admit it is not an easy problem to solve at all (and really appreciate, there are people like you who do not run away from that mess and keep searching for a solution).

After a bit more testing (only in `eu-west-1` region) I have found, that:

- using both decorators does not help (both dotted and non-dotted bucket names were failing)
- using only `@mock_s3bucket_path` helps for dotted bucket names and fails on non-dotted bucket names
- as shown before, using only `@mock_s3bucket_path` fails on dotted bucket names.

My current conclusion is: I have to use specific mockup according to names of bucket in question (with other exceptional situation in `eu-central-1` region - but that I am trying to avoid).

Merged (and functional) decorator would be really welcome.

A side note:

With AWS S3 API (boto3) I am rather disappointed, that while bucket names are globally valid (across all regions), top level library like `boto3` requires user to care about what region the bucket lives in. Described in issue [S3 buckets are global and shall not bother clients with region_name after creation](boto/boto3#145). Anyway, this is problem with `boto3` not with `moto`, which only has to deal with growing complexity of situations to face.
